### PR TITLE
[BEAM-3516] Spanner BatchFn does not respect mutation limits

### PR DESCRIPTION
**Pull request by NathanHowell**
Estimate the number of mutations in in a group by counting the affected columns and associated indexes.

The code currently assumes each row contains a single mutation and has flushes batches _after_ a (hardcoded) 10k row threshold has been exeeded. Spanner rejects commits that exceed 20k mutations, including indexes. This disconnect between the estimated mutations and the actual count causes commit failures.

This change estimates the actual mutations by counting the number of indexes covering each column, and summing up the counts of columns and indexes contained within a MutationGroup. The group is flushed prior to the limit being exceeded.

### CODE REVIEW

**Comment by NathanHowell** [comment]
Hi @mairbek and @dhalperi, could you take a look at this change? It's a bit light on tests..

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/MutationCellEstimator.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        // for single keys simply sum up all the columns in the schema
+        final long rows = Iterables.size(keySet.getKeys());
+        for (long cells : columnCells.values()) {
+          mutatedCells += rows * cells;
+        }
+
+        // ranges should already be broken up into individual batches
+        // but just in case, make a worst-case estimate about the size
+        // of the key range so they will get their own transaction
+        final long ranges = Iterables.size(keySet.getRanges());
```
Comment (mairbek) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/MutationCellEstimator.java]: We only batch single key deletes. I think we can return zero or -1 here and avoid passing `maxNumMutations`

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/ReadSpannerSchema.java [hunk 2]
```diff
@@ -76,11 +77,28 @@ public void processElement(ProcessContext c) throws Exception {
...
+            + "    c.table_name"
+            + "  , c.column_name"
+            + "  , c.spanner_type"
+            + "  , (1 + COALESCE(t.indices, 0)) AS cells_mutated"
+            + "  FROM ("
+            + "    SELECT c.table_name, c.column_name, c.spanner_type, c.ordinal_position"
+            + "     FROM information_schema.columns as c"
+            + "     WHERE c.table_catalog = '' AND c.table_schema = '') AS c"
+            + "  LEFT OUTER JOIN ("
+            + "    SELECT t.table_name, t.column_name, COUNT(1) AS indices"
```
Comment (mairbek) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/ReadSpannerSchema.java]: Could we use COUNT(*) here? I think it is more common.

**Comment by dhalperi** [comment]
Given current job responsiibilities, will not be able to review in a reasonable amount of time -- sorry!

---

Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/apache/beam/4860#-:-L887s-l1nrLCmTkeK4c:b-hl6pcc)*

**Comment by NathanHowell** [comment]
@mairbek I've made the requested changes, can you take another look? (an unrelated test is failing atm)

**Comment by mairbek** [comment]
lgtm, @chamikaramj can you take a look?

**Comment by mairbek** [comment]
retest this please

**Comment by mairbek** [comment]
I'm looking to make it part of Beam 2.5.0

**Comment by NathanHowell** [comment]
Excellent. We're no longer using Spanner (AWS won the contract) so I can't do much testing... but I can probably get it rebased against master later this week.

**Comment by jkff** [comment]
This PR needs to be rebased. Please also take a look at precommit failures and fix them.

**Comment by NathanHowell** [comment]
Fixed in #5297, thanks for the help! 👍

**Pull request closed without merging**